### PR TITLE
feat: support deepseek model in camel

### LINKED ISSUE #1199: [Feature Request] Support `deepseek` models
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [ ] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

Support `deepseek` models

### Solution

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by NeilJohnson0930**
## Description

Describe your changes in detail.

## Motivation and Context

`close #1199 ` 

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/configs/deepseek_config.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+from camel.types import NOT_GIVEN, NotGiven
+
+
+class DeepSeekConfig(BaseConfig):
+    r"""Defines the parameters for generating chat completions using the
+    OpenAI API.
+
+    Args:
+        temperature (float, optional): Possible values: <= 2
+            Default value: 1
```
Comment (Wendong-Fan) [inline camel/configs/deepseek_config.py:30]: doc string format could be polished, could you refer to https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md#3-document-parameters-in-the-args-section or the doc string in openai config?

File: camel/models/deepseek_model.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    @property
+    def token_counter(self) -> BaseTokenCounter:
+        r"""Initialize the token counter for the model backend.
+
+        Returns:
+            BaseTokenCounter: The token counter following the model's
+                tokenization style.
+        """
+        if not self._token_counter:
+            self._token_counter = OpenAITokenCounter(self.model_type)
```
Comment (Wendong-Fan) [inline camel/models/deepseek_model.py:83]: OpenAITokenCounter could not accept deepseek model, use `ModelType.GPT_4O_MINI`

File: camel/models/deepseek_model.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        Args:
+            messages (List[OpenAIMessage]): Message list with the chat history
+                in DeepSeek API format.
+
+        Returns:
+            Union[ChatCompletion, Stream[ChatCompletionChunk]]:
+                `ChatCompletion` in the non-stream mode, or
+                `Stream[ChatCompletionChunk]` in the stream mode.
+        """
+        model_type_str = UnifiedModelType.__str__(self.model_type)
```
Comment (Wendong-Fan) [inline camel/models/deepseek_model.py:102]: i think it's not necessary

File: test/models/test_deepseek_model.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+"""
+please set the below os environment:
+export DEEPSEEK_BASE_URL=""
+export DEEPSEEK_API_KEY=""
```
Comment (Wendong-Fan) [inline test/models/test_deepseek_model.py:17]: this test doesn't require key setting

File: camel/configs/deepseek_config.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            are present. "auto" is the default if tools are present.
+            Defaults to "auto".
+        logprobs (bool, optional): Whether to return log probabilities of
+            the output tokens or not. If true, returns the log probabilities
+            of each output token returned in the content of message.
+            Defaults to False.
+        top_logprobs (int, optional): An integer between 0 and 20 specifying
+            the number of most likely tokens to return at each token
+            position, each with an associated log probability. logprobs
+            must be set to true if this parameter is used. Defaults to None.
```
Comment (Wendong-Fan) [inline camel/configs/deepseek_config.py:101]: seems not set

File: camel/configs/deepseek_config.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+class DeepSeekConfig(BaseConfig):
+    r"""Defines the parameters for generating chat completions using the
+    DeepSeek API.
+
+    Args:
+        temperature (float, optional): The sampling temperature to use,
+            between 0 and 2. Higher values like 0.8 will make the output
+            more random, while lower values like 0.2 will make it more
+            focused and deterministic. We generally recommend altering
+            this or top_p but not both. Defaults to 1.
```
Comment (Wendong-Fan) [inline camel/configs/deepseek_config.py:33]:
```suggestion
        temperature (float, optional): Sampling temperature to use, between
            :obj:`0` and :obj:`2`. Higher values make the output more random,
            while lower values make it more focused and deterministic.
            (default: :obj:`0.2`)
```

File: camel/configs/deepseek_config.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            they become available, with the stream terminated by a
+            data: [DONE] message. Defaults to False.
+        stream_options (bool, optional): Options for streaming response.
+            Only set this when you set stream: true. include_usage
+            (boolean) If set, an additional chunk will be streamed before
+            the data: [DONE] message. The usage field on this chunk shows
+            the token usage statistics for the entire request, and the
+            choices field will always be an empty array. All other chunks
+            will also include a usage field, but with a null value.
+            Defaults to False.
```
Comment (Wendong-Fan) [inline camel/configs/deepseek_config.py:64]: seems not set

**Review by Wendong-Fan — COMMENTED** [review]
Thanks @NeilJohnson0930 , could you also update docs/key_modules/models.md add deepseek to the list?

**Comment by MuggleJinx** [comment]
Hi @NeilJohnson0930, thanks for the PR. However, I ran into some problem when running DeepSeek_model_example.py, 
seems there is some configuration issue, maybe need to modify the `stream` or `stream_option` in the deepseek_config file. Could you check this? Thanks!

```
File "/Users/jinx0a/micromamba/envs/ai/lib/python3.12/site-packages/openai/_base_client.py", line 1058, in _request
    raise self._make_status_error_from_response(err.response) from None
openai.UnprocessableEntityError: Failed to deserialize the JSON body into the target type: stream_options: invalid type: boolean `false`, expected struct StreamOption at line 1 column 361
```

File: camel/configs/deepseek_config.py [hunk 8]
```diff
@@ -110,6 +109,9 @@ class DeepSeekConfig(BaseConfig):
     response_format: Union[Type[BaseModel], dict, NotGiven] = NOT_GIVEN
     frequency_penalty: float = 0.0
     tool_choice: Optional[Union[dict[str, str], str]] = None
+    stream_options: bool = False
```
Comment (MuggleJinx) [inline camel/configs/deepseek_config.py:112]: Here!

**Review by Wendong-Fan — APPROVED** [review]
Thanks @NeilJohnson0930 , added one more commit here: https://github.com/camel-ai/camel/pull/1217/commits/da445eb326c8143d42ba3002efd92a474e242d1f, now it looks great, thanks a lot !

File: camel/configs/deepseek_config.py (hunk 8 above)
Comment (NeilJohnson0930) [reply camel/configs/deepseek_config.py]: got it, problem solved

**Comment by NeilJohnson0930** [comment]
> Hi @NeilJohnson0930, thanks for the PR. However, I ran into some problem when running DeepSeek_model_example.py, seems there is some configuration issue, I think you need to modify the `stream_option` in the deepseek_config file. Could you check this? Thanks!
> 
> ```
> File "/Users/jinx0a/micromamba/envs/ai/lib/python3.12/site-packages/openai/_base_client.py", line 1058, in _request
>     raise self._make_status_error_from_response(err.response) from None
> openai.UnprocessableEntityError: Failed to deserialize the JSON body into the target type: stream_options: invalid type: boolean `false`, expected struct StreamOption at line 1 column 361
> ```
Thanks MuggleJinx,  config modification has been done.

**Comment by MuggleJinx** [comment]
Thanks, looks cool to me!

**Pull request merged**